kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportMapper.kt

```diff
@@ -75,8 +75,15 @@ private fun isSealedClassConstructor(descriptor: ConstructorDescriptor) = descri
 
 // Note: partially duplicated in ObjCExportLazyImpl.translateTopLevels.
 internal fun ObjCExportMapper.shouldBeExposed(descriptor: CallableMemberDescriptor): Boolean =
-        descriptor.isEffectivelyPublicApi && !descriptor.isExpect &&
-                !isHiddenByDeprecation(descriptor) && !(descriptor is ConstructorDescriptor && isSealedClassConstructor(descriptor))
+        descriptor.isEffectivelyPublicApi && !descriptor.isExpect && !isHiddenByDeprecation(descriptor) &&
+                !(descriptor is ConstructorDescriptor && isSealedClassConstructor(descriptor)) && !descriptor.isObjCRefined()
+
+private fun CallableMemberDescriptor.isObjCRefined(): Boolean {
+    if (overriddenDescriptors.isNotEmpty()) return overriddenDescriptors.all { it.isObjCRefined() }
```

Comment:
Should the compiler explicitly prohibit using `@ObjCRefined` and `@SwiftRefined` on `override` functions? Otherwise, they seem silently ignored.

Reply:
Yeah they are currently ignored since the goal of the annotations is to refine the public API.
It wouldn't really make sense to refine an `override` function but not the base one, right? 

Is it possible for the compiler to prohibit applying these annotations on overrides?

Reply:
> It wouldn't really make sense to refine an `override` function but not the base one, right? 

Well, I can imagine a case like:
```kotlin
interface ThisTypeIsGreatForSwift
interface ThisTypeIsUglyInSwift : ThisTypeIsGreatForSwift

interface Base {
    fun foo(): ThisTypeIsGreatForSwift
}

abstract class Derived : Base {
    @SwiftRefined override fun foo(): ThisTypeIsUglyInSwift
}
```

> Is it possible for the compiler to prohibit applying these annotations on overrides?

Yes. A good way to do this is to implement a frontend checker; a nice example can be found at https://github.com/JetBrains/kotlin/blob/4f9364f997d240345d343cc8b6e4c00c57ddd878/native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeThreadLocalChecker.kt#L32.

Reply:
I think you can still solve that use case without annotation the override.
Since `ThisTypeIsGreatForSwift` is great for Swift that would imply that some function or property in `ThisTypeIsUglyInSwift` is ugly in Swift. Which means you would annotate that particular function or property.

However now that I am thinking about this some more I guess it could make sense.
```kotlin
interface Base {
    @SwiftRefined fun foo(): List<String>
}

interface Derived: Base {
    @SwiftRefined override fun foo(): MutableList<String>
}
```

> Yes. A good way to do this is to implement a frontend checker; 

Nice! Will take a look at that.

Reply:
> Since `ThisTypeIsGreatForSwift` is great for Swift that would imply that some function or property in `ThisTypeIsUglyInSwift` is ugly in Swift. Which means you would annotate that particular function or property.

Not always -- "ugliness" might be the property of a type itself, its supertypes etc. Also, what if this type is from a third-party library?

I'm not proposing to support this case, just trying to explore it :)

> However now that I am thinking about this some more I guess it could make sense.

On the other hand, supporting it might be complicated. Generally, ObjCExport is not ready for the case when a function is not exposed, but it overrides and it is overridden by exposed functions. So maybe it is better to prohibit using `@*Refined` on `override` functions. We will always be able to allow this later if we find this really important.

Reply:
> Do you have an example? And how would a refined class behave?:

```kotlin
interface ThisTypeIsGreatForSwift
value class ThisTypeIsUglyInSwift(val value: String) : ThisTypeIsGreatForSwift

interface Base {
    fun foo(): ThisTypeIsGreatForSwift
}

abstract class Derived : Base {
    override abstract fun foo(): ThisTypeIsUglyInSwift
}
```

The last `foo` method just gets `id` as a return type ([KT-32352](https://youtrack.jetbrains.com/issue/KT-32352/inline-classes-are-exported-as-id-to-objc-swift)).

> In the above example `UglyInterface` shouldn't be exposed to ObjC.
> So what would happen to the `getUglyInterface` method? How should you define the refined type/method?

The closest thing we have at the moment is https://github.com/JetBrains/kotlin/blob/b5933c70e20a0dfd788d8d4a6c04ca0d16f7c731/kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt#L898

> IMO it would be the responsibility of the library to include refined APIs.

In practice, it doesn't work this way.

> Clients shouldn't be able to change the API of a library, right?

The original example doesn't change the API of a library, it changes only our own API.

> I think the annotations should be inherited. So a function that overrides a function with a `@*Refined` is also refined.

Right.

> That would make annotating the override function useless.

No, see the original example where we have to refine the overriding function but not the overridden one.

> It should however still be possible to use a meta annotated annotation on the override function.
That way the above example with the `List` and `MutableList` would still allow an annotation processor to generate a refined API for the override function.

What do you mean, could you provide an example?

Reply:
> The last `foo` method just gets `id` as a return type ([KT-32352](https://youtrack.jetbrains.com/issue/KT-32352/inline-classes-are-exported-as-id-to-objc-swift)).
> The closest thing we have at the moment is

So what would be the point of `@ObjCRefined` on a class if it results in the type being represented as `id`?

> In practice, it doesn't work this way.
> The original example doesn't change the API of a library, it changes only our own API.

True. But that means we change the functions/properties in our own API that use the library class.
Meaning you would have to annotate the client API:
```kotlin
// library
interface SomeUglyInObjCLibraryInterface

// client
interface SomeClientInterface {
    @ObjCRefined
    fun getLibraryInterface(): SomeUglyInObjCLibraryInterface
}

interface SomeNotUglyInObjCClientInterface

fun SomeClientInterface.getLibraryInterfaceRefined(): SomeNotUglyInObjCClientInterface = TODO()
```

> No, see the original example where we have to refine the overriding function but not the overridden one.

Hmm right. So what do you mean exactly with:
> ObjCExport is not ready for the case when a function is not exposed, but it overrides and it is overridden by exposed functions.

> What do you mean, could you provide an example?

```kotlin 
// annotation processor
@SwiftRefinedMeta
annotation class RefineLists

// client
interface Base {
    @RefineLists fun foo(): List<String>
}

interface Derived: Base {
    @RefineLists override fun foo(): MutableList<String>
}

// generated code
fun Base.fooRefined(): RefinedList<String> = TODO()
fun Derived.fooRefined(): RefinedMutableList<String> = TODO()
```

The `@RefineLists` annotation should be allowed on the `Derived.foo` function such that it can generated the `Derived.fooRefined` function.

Reply:
> So what would be the point of `@ObjCRefined` on a class if it results in the type being represented as `id`?

There is no point in that. Why do you ask? Did I propose this approach?

> True. But that means we change the functions/properties in our own API that use the library class.
Meaning you would have to annotate the client API:

Your example doesn't have `override`, my [original example](https://github.com/JetBrains/kotlin/pull/4818#discussion_r876025877) does. That's what makes it harder, and that's what requires thorough design.

> Hmm right. So what do you mean exactly with:
> > ObjCExport is not ready for the case when a function is not exposed, but it overrides and it is overridden by exposed functions.

Consider the example:
```kotlin
open class A {
    open fun foo() {}
}

open class B : A() {
    override fun foo() {}
}

open class C : B() {
    override fun foo() {}
}
```

Imagine it is allowed to have `shouldBeExposed` return `true` for `A.foo` and `C.foo` but `false` for `B.foo`. ObjCExport is not ready for that -- some things might go wrong.
So, if you allow having `@ObjCRefined` on an `override` method but not on a method it overrides, then you have to also hide all methods that override the refined method.
Moreover, I don't think it is safe to make `shouldBeExposed` return `false` for a method if it overrides a method that has `shouldBeExposed == true`. So, generally, refinement should be inherited. See also https://github.com/JetBrains/kotlin/blob/4dcfd3823664d765e62b33020623c9487ab56abb/kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportMapper.kt#L85.

> The `@RefineLists` annotation should be allowed on the `Derived.foo` function such that it can generated the `Derived.fooRefined` function.

Got you, thank you for the explanation. Makes sense.

Reply:
> There is no point in that. Why do you ask? Did I propose this approach?

Well you [mentioned](https://github.com/JetBrains/kotlin/pull/4818#discussion_r876671433) that "ugliness" might be the property of a type itself.
I am trying to understand what that would mean. And particularly how you would refine an "ugly" type.

> So, generally, refinement should be inherited.

Alright. So that means you can only annotate declarations with the `@*Refined` annotations if they are Kotlin "base" declarations. Annotating an override declaration is only allowed if all of the overridden declarations are also annotated.

I guess that's fine since only a small number of use cases would require the refinement of an override declarations (if the overridden ones aren't refined).

Reply:
> Well you [mentioned](https://github.com/JetBrains/kotlin/pull/4818#discussion_r876671433) that "ugliness" might be the property of a type itself.
I am trying to understand what that would mean.

I basically meant types that are exported to Swift or Objective-C in an undesirable way, like value types in my example above.

> And particularly how you would refine an "ugly" type.

As I said, this is not always an option.

> Alright. So that means you can only annotate declarations with the `@*Refined` annotations if they are Kotlin "base" declarations. Annotating an override declaration is only allowed if all of the overridden declarations are also annotated.

Mostly for `ObjCRefined`, I guess.
Yes, this should work. There are alternatives -- like reporting warnings for such cases.